academy/0.0-B9lab-Blockchains/3_managed.md

```diff
@@ -0,0 +1,134 @@
+
+In simple terms, the halting problem describes a scenario in which a program loops forever. Ethereum's solution is to introduce **Gas** as a fee for each computational step. Every block has a maximum Gas limit which limits the number of computational steps that can be executed per block - just like a combustion engine, if the Gas runs out then the program stops. Gas only exists during the execution of a transaction and the user is free to specify any **Gas price** in terms of **Ether**, a value which Ethereum requires to convert **Ether to Gas**.
+
+<HighlightBox type="info">
+
+**Ether** is the **native cryptocurrency** of the Ethereum platform, as Bitcoin is the native cryptocurrency of the Bitcoin blockchain.
+
+</HighlightBox>
+
+The protocol defines the **Gas cost** for each operation in the EVM. The halting problem is resolved by attaching financial costs to every step in a program and by limiting the funds available. Transactions that do not compute successfully and are not on budget are rejected.
+
+In Ethereum, **smart contracts** are stored on every single node - they are public. In practice, a smart contract in the EVM is an autonomous agent with an internal account. The most popular language for writing such a contract is **Solidity**.
+
+<ExpansionPanel title="Faster block time">
+
+Ethereum has a much faster block time than Bitcoin - currently it is around 15 seconds. This is made possible thanks to Ethereum's implementation of the [GHOST protocol](http://www.cs.huji.ac.il/~yoni_sompo/pubs/15/inclusive_full.pdf). Ethereum implements a variation of Bitcoin's PoW consensus algorithm called Ethash, which is intended to be [ASIC-resistant and GPU friendly](https://github.com/ethereum/wiki/wiki/Ethash-Design-Rationale). Core developers of the Ethereum project plan to switch to the **proof-of-stake (PoS)** consensus algorithm in the future.
+
+A well-understood challenge related to reduced block time stems from network latency. Since nodes do not learn about newly discovered blocks simultaneously, at any given time a portion of the miners will be working on already solved (i.e. old) blocks. If these miners find a solution, they might not be rewarded and a "fork" of the chain results, in which rival chains with a common true history spread through the network. The rate at which such transient forks occur increases predictably with shorter block times. While Bitcoin addresses this efficiency issue with a relatively long block time of 10 minutes, Ethereum addresses this concern with a partial reward strategy: valid blocks that aren't ultimately included in the canonical chain (because another chain becomes longer) are still included on the side and known as **uncles**. Miners of uncles receive a smaller reward than regular block miners. 
+
+![Uncles in Ethereum](images/00_18_block_uncles-01.png)
```

Comment:
Not directly a "language" review factor, but: as a layperson, I find the depiction of the blockchain/uncles dynamic in the graphic _00_18_block_uncles-01.png_ a bit counter-intuitive. It seems to me that the direction of the arrows AND the atypical right-to-left orientation to represent old-to-new could provoke confusion.

I would clarify by expanding the labels "Block" and "Uncle" to indicate newness, and orient the "aging" of the system to show older at the left and newer at the right (see attached image).

![Blocks-and-uncles](https://user-images.githubusercontent.com/104000630/165515303-f1ac5725-4d8b-4c7e-b0ee-326ccdf13f5c.png)

I have maintained the use of arrows which point towards the source of a block's history. However, I would observe that this also seems counter-intuitive, with diagrams typically pointing _from_ one state _toward_ a subsequent state. Perhaps the usage here is a convention in blockchain specifically, but if not I would be inclined to reverse the arrow directions in all instances here.

Reply:
It is an issue with which we often grapple. Personally, I would add a _passage of time_ arrow below all going from left to right, the `x` axis. Perhaps with the legend of "time of appearance on network" or suchlike.
Then have each block put alongside the `x` axis as is visible in the 2 versions of the diagram.
The arrows would point leftward, but I would make them tenuous, or dotted, with a legend of: A <- B: A is parent block of B.
And name _Uncle B_ a block that appears later than _Block B_ but is then included in _Block C_, not _Block A_. The arrow would point from _Block C_ to _Uncle B_ with another tenuous arrow type and the legend A <- B: A is an uncle of B.
Keep in mind that this uncle concept is non-existent in PoS blockchains. So we could just mention it in passing with a link.
@rob-Hitchens

Reply:
I think the passage of time arrow is a good idea to possibly add in the future but not crucial for this stage. We've used this image before and it's just in passing mention, in this case, to help position Cosmos against the competition. The complexity of this process is one of the elements Tendermint (arguably) does better with validator sets and true transaction finality.